### PR TITLE
Add Timer unit tests

### PR DESCRIPTION
**Pull request by 2nafish117**
Add timer tests
link to #43440

### CODE REVIEW

**Comment by 2nafish117** [comment]
The Timer start/stop test case is currently failing because it is not part of a scene tree, is there a test scene tree that i need to add it to, or should i make a scene tree in cpp and add it to it?

**Comment by Chaosus** [comment]
Thanks for your PR, CI failed, use the clang-format tool to prevent such errors (just an advice) (see https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#using-clang-format-locally). Also, commits need to be squashed (see https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#modifying-a-pull-request).

**Comment by 2nafish117** [comment]
> Thanks for your PR, CI failed, use the clang-format tool to prevent such errors (just an advice) (see https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#using-clang-format-locally). Also, commits need to be squashed (see https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#modifying-a-pull-request).

@Chaosus ive managed to fix the clangfmt issue and squashed the commits, the only thing that remains is to fix the start/stop tests requiring a scene tree

File: tests/scene/test_timer.h
Comment (RedMser) [inline tests/scene/test_timer.h:109]: I've seen the Node tests do it like this:
```cpp
SceneTree::get_singleton()->get_root()->add_child(node);
```

**Comment by RedMser** [comment]
You might want to test the main functionality of timer: the timeout signal. But do so with a short wait time, so it does not take long for the test to complete:

* Measure current time using Time singleton.
* Start a one-shot timer with wait time 0.1 seconds.
* On timeout, measure time again using Time singleton.
* Ensure the difference between timestamps is at least 0.1 seconds (might be longer, but never shorter).

Could also check e.g. if timer is no longer running after timeout (or if it's not a one-shot timer, ensure that it is still running).

File: tests/scene/test_timer.h
Comment (2nafish117) [reply tests/scene/test_timer.h:109]: Ah, thank you.

**Comment by 2nafish117** [comment]
> You might want to test the main functionality of timer: the timeout signal. But do so with a short wait time, so it does not take long for the test to complete:
> 
>     * Measure current time using Time singleton.
> 
>     * Start a one-shot timer with wait time 0.1 seconds.
> 
>     * On timeout, measure time again using Time singleton.
> 
>     * Ensure the difference between timestamps is at least 0.1 seconds (might be longer, but never shorter).
> 
> 
> Could also check e.g. if timer is no longer running after timeout (or if it's not a one-shot timer, ensure that it is still running).

to update the timer, do i also have to manually process the scene tree like so?

SceneTree::get_singleton()->process(0.1);
SceneTree::get_singleton()->physics_process(0.1);

**Comment by RedMser** [comment]
> to update the timer, do i also have to manually process the scene tree like so?

I do see calls to that function in the Node tests, so yeah I think so. Also you might want to test process and physics_process separately, depending on which process callback the Timer is set to (so only tick one of the two, not both at once).

**Comment by 2nafish117** [comment]
@RedMser I tried getting the timeout signal to be called by manually calling process on the scene tree, but the callback doesnt seem to be getting called. Do you know why that might be?

File: tests/scene/test_timer.h
Comment (RedMser) [inline tests/scene/test_timer.h:167]: Not sure how to fix, but maybe you need to memnew this struct as well?

**Review by RedMser — COMMENTED** [review]
I never wrote proper unit tests for Godot before, so I'm trying to figure it out based on looking at other tests haha. Maybe someone with more of an idea can help you out, e.g. in the godot developers chat...

File: tests/scene/test_timer.h
Comment (2nafish117) [reply tests/scene/test_timer.h:167]: I believe it isnt possible to test the signals of nodes in unit tests, i dont see any other node signals being tested.
Ill ask in the dev chat as well, thanks.

Comment (2nafish117) [reply tests/scene/test_timer.h]: it is possible to test the signals with the SIGNAL_WATCH and SIGNAL_CHECK macros.

**Comment by 2nafish117** [comment]
@RedMser ive rebased this with latest from master, could you have a look at this PR once again.

**Review by RedMser — APPROVED** [review]
Seems fine, but you should wait for a maintainer to approve the workflow run, so you can see if the tests succeed in the CI/CD as well. Maybe you can ask in the [Godot Contributors chat](https://chat.godotengine.org/), if you don't want to wait.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Pull request merged**
